### PR TITLE
[c10d] implement init_process_group in C++

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45930 [c10d] implement init_process_group in C++**
* #46411 [c10d] add c10d_frontend to build, add getter/setter

### CODE REVIEW

File: torch/csrc/distributed/c10d/c10d_frontend.cpp [hunk 1]
```diff
@@ -1,73 +1,147 @@
...
+
+  auto it = std::find_if(
+      pg_names_.begin(),
+      pg_names_.end(),
+      [&](const std::pair<std::shared_ptr<ProcessGroup>, std::string>&
+              pg_name) { return pg_name.second == *group_name; });
+
+  if (it == pg_names_.end()) {
+    throw std::runtime_error(
+        "The specified group name has already been "
```
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/c10d_frontend.cpp:67]: Nit: Can we specify the name that's been taken to help debugging?

File: torch/csrc/distributed/c10d/c10d_frontend.cpp [hunk 2]
```diff
@@ -1,73 +1,147 @@
...
+#endif
+    } else if (backend == "nccl") {
+#ifdef USE_C10D_NCCL
+      auto options = ProcessGroupNCCL::Options();
+
+      options.isHighPriorityStream = false;
+      options.opTimeout = timeout;
+      pg = std::make_shared<ProcessGroupNCCL>(
+          prefix_store, rank, world_size, options);
+#endif
```
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/c10d_frontend.cpp:129]: What error do we throw if not compiled with `USE_C10D_NCCL` and the user requests NCCL backend?

File: torch/csrc/distributed/c10d/c10d_frontend.cpp [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+#ifdef USE_C10D_MPI
+#include <c10d/ProcessGroupMPI.hpp>
+#endif
+
+namespace c10d {
+
+const std::string& DistributedC10d::backend() const {
+  return backend_;
+}
+void DistributedC10d::backend(std::string const& backend_name) {
```
Comment (gmagogsfm) [inline torch/csrc/distributed/c10d/c10d_frontend.cpp:25]: Nit: setBackend? same for other setters.

File: torch/csrc/distributed/c10d/c10d_frontend.cpp [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    ++group_count_;
+  }
+
+  auto it = std::find_if(
+      pg_names_.begin(),
+      pg_names_.end(),
+      [&](const std::pair<std::shared_ptr<ProcessGroup>, std::string>&
+              pg_name) { return pg_name.second == *group_name; });
+
+  if (it == pg_names_.end()) {
```
Comment (gmagogsfm) [inline torch/csrc/distributed/c10d/c10d_frontend.cpp:65]: Should this be `it != pg_names_.end()`?

File: torch/csrc/distributed/c10d/c10d_frontend.cpp [hunk 5]
```diff
@@ -31,4 +48,105 @@ void DistributedC10d::set_default_pg_init_method(
...
+std::shared_ptr<ProcessGroup> DistributedC10d::newProcessGroupHelper(
+    const int64_t world_size,
+    const int64_t rank,
+    const std::vector<int64_t>& group_ranks,
+    const std::string& backend_str,
+    const std::shared_ptr<Store>& store,
+    c10::optional<std::string> group_name,
+    std::chrono::milliseconds timeout) {
+  if (!group_name.has_value()) {
+    group_name = std::to_string(group_count_);
```
Comment (pritamdamania87) [inline torch/csrc/distributed/c10d/c10d_frontend.cpp:60]: nit: `group_name = std::to_string(group_count_++)`;

File: torch/csrc/distributed/c10d/c10d_frontend.cpp [hunk 6]
```diff
@@ -31,4 +48,105 @@ void DistributedC10d::set_default_pg_init_method(
...
+    throw std::runtime_error(
+        "The specified group name has already been "
+        "created, please use a different group name");
+  }
+
+  bool is_default_group = pg_group_ranks_.size() == 0;
+
+  std::shared_ptr<ProcessGroup> pg = nullptr;
+
+  std::string backend = Backend::get(backend_str);
```
Comment (pritamdamania87) [inline torch/csrc/distributed/c10d/c10d_frontend.cpp:80]: Where is this API implemented?

File: torch/csrc/distributed/c10d/c10d_frontend.cpp [hunk 7]
```diff
@@ -31,4 +48,105 @@ void DistributedC10d::set_default_pg_init_method(
...
+  if (backend == "mpi") {
+#ifdef USE_C10D_MPI
+    pg = ProcessGruopMPI::createProcessGroupMPI(group_ranks);
```
Comment (pritamdamania87) [inline torch/csrc/distributed/c10d/c10d_frontend.cpp:83]: s/ProcessGruopMPI/ProcessGroupMPI/, this code wouldn't compile?

File: torch/csrc/distributed/c10d/c10d_frontend.cpp [hunk 8]
```diff
@@ -31,4 +48,105 @@ void DistributedC10d::set_default_pg_init_method(
...
+#else
+    throw std::runtime_error(
+        "Distributed package doesn't have MPI built in."
+        " MPI is only included if you build PyTorch from"
+        " source on a host that has MPI installed.");
+#endif
+  } else {
+    if (!is_default_group) {
+      int64_t global_rank = default_pg_->getRank();
+      if (std::find(group_ranks.begin(), group_ranks.end(), global_rank) ==
```
Comment (pritamdamania87) [inline torch/csrc/distributed/c10d/c10d_frontend.cpp:93]: Can we retain the comments from distributed_c10d.py for clarity for developers?

File: torch/csrc/distributed/c10d/c10d_frontend.cpp [hunk 9]
```diff
@@ -31,4 +48,105 @@ void DistributedC10d::set_default_pg_init_method(
   default_pg_init_method_ = init_method;
 }
 
+std::shared_ptr<ProcessGroup> DistributedC10d::newProcessGroupHelper(
+    const int64_t world_size,
+    const int64_t rank,
+    const std::vector<int64_t>& group_ranks,
```
Comment (pritamdamania87) [inline torch/csrc/distributed/c10d/c10d_frontend.cpp:54]: Shouldn't we have some unit tests covering this new API? My proposal would be to actually modify our current code in `distributed_c10d.py` to use this and as a result all of our existing tests would cover it.

Comment (gmagogsfm) [reply torch/csrc/distributed/c10d/c10d_frontend.cpp:54]: Hi Pritam, great point! Soon, we plan to reuse all existing tests that involve torch.distributed by adding a test configuration that toggles which c10d frontend the test uses. This way we ensure the two frontends remain consistent. That's also why we don't have new test cases just for new implementation for now.

**Pull request closed without merging**